### ISSUE TITLE
Add chunked implementation of siglip sigmoid loss

### ISSUE THREAD

**Issue by cdeepakroy** [google-research/big_vision#130]
Thank you for the great work on siglip paper. In particular, the gains in metrics at small batch sizes are impressive.
[Sigmoid Loss for Language Image Pre-Training](https://arxiv.org/abs/2303.15343)

I tried using this pytorch based distributed chunked implementation in open clip repo using torch.distributed.P2POp
https://github.com/mlfoundations/open_clip/blob/fc5a37b72d705f760ebbc7915b84729816ed471f/src/open_clip/loss.py#L307

I tried taking a ViT B vision encoder + XLM Roberta text encoder and train it using both CLIP softmax and SigLip sigmoid loss on an in house dataset of 10M image-text pairs at an effective batch size of 9k (with V100 GPUs) and observed that CLIP softmax still performs better than siglip sigmoid loss on nDCG metric. 

I was wondering if there is any error in above implementation using p2pop. I also tried using an all_gather to get negative text_features from other gpus but still the behavior seems to be the same

```
    def forward(self, image_features, text_features):
        
        loss = self._loss(image_features, text_features, self.logit_scale, self.logit_bias)

        if global_manager.world_size > 1:

            # Gather text features from all ranks
            text_features_dict = {'text_features': text_features, 'global_rank': global_manager.rank}
            all_text_features_dict = all_gather_objects(text_features_dict)

            # Compute loss against negative text features from all other ranks
            for i in range(len(all_text_features_dict)):
                neigh_rank = all_text_features_dict[i]['global_rank']
                neigh_text_features = all_text_features_dict[i]['text_features']
                if neigh_rank != global_manager.rank:
                    loss += self._loss(image_features, neigh_text_features, self.logit_scale, self.logit_bias, negative_only=True)
                    
        return {"loss": loss}
```

The implementation in this repo seems to be the non-chunked version
https://github.com/google-research/big_vision/blob/46b2456f54b9d4f829d1925b78943372b376153d/big_vision/trainers/proj/image_text/siglip.py#L287

I am wondering if you can add a chunked implementation of the siglip sigmoig loss.

**Comment by avocardio** [comment]
I implemented the chunked sigmoid loss that the TODO in `siglip.py` mentions. Here's a pure JAX version using `lax.scan` + `custom_vjp` — no external dependencies, works on TPU and GPU.

**Memory: O(B × chunk_size) instead of O(B²).** At B=32768 with chunk=512, that's 576MB vs 16.7GB for logits alone.

The key trick: `lax.scan` with `unroll=4` eliminates the loop overhead, matching reference speed while chunking the computation:

```python
@partial(jax.custom_vjp, nondiff_argnums=(4,))
def chunked_siglip_loss(img_features, txt_features, temperature, bias, chunk_size=512):
    B, D = img_features.shape
    chunk_size = min(chunk_size, B)

    def scan_body(loss_acc, chunk_start):
        txt_chunk = jax.lax.dynamic_slice(txt_features, (chunk_start, 0), (chunk_size, D))
        raw_logits = jnp.dot(img_features, txt_chunk.T)
        logits = raw_logits * temperature + bias
        img_idx = jnp.arange(B)[:, None]
        txt_idx = chunk_start + jnp.arange(chunk_size)[None, :]
        targets = (img_idx == txt_idx).astype(logits.dtype)
        bce = jnp.maximum(logits, 0) - logits * targets + jnp.log1p(jnp.exp(-jnp.abs(logits)))
        return loss_acc + jnp.sum(bce), None

    num_chunks = B // chunk_size
    chunk_starts = jnp.arange(num_chunks) * chunk_size
    total_loss, _ = jax.lax.scan(scan_body, jnp.float32(0.0), chunk_starts, unroll=min(num_chunks, 4))
    return total_loss / B
```

The backward pass is also chunked via `custom_vjp` (never materializes B² gradients). Full implementation with tests: https://github.com/avocardio/siglip-kernel/blob/main/jax/chunked_loss.py

Benchmarked on H200 (141GB):
| B | Reference | Chunked | Memory savings |
|---|----------|---------|----------------|
| 16384 | 33ms | 33ms (1.0x) | 14.7x less |
| 32768 | 131ms | 131ms (1.0x) | 29x less |
| 65536 | 513ms | 513ms (1.0x) | 57x less |

For the distributed cross-device version (`shard_map`), the same pattern extends — iterate over remote text feature shards via ring exchange, accumulating loss per chunk. Happy to share that implementation too if there's interest.
